### PR TITLE
Refactoring of search

### LINKED ISSUE #212: Document or simplify search index entry construction
Document and/or simplify the construction of search index entries.

### LINKED ISSUE #213: Move search snippet generation to backend service
The snippets (the properties that contain the search-string) are currently constructed in the frontend, while this should be part of the backend service and the snippet should be included in the search result. Move the search snippet generation to the backend service and include the snippet in the search-result.

### PR DESCRIPTION
**Pull request by CatoLeanTruetschel**
Closes #212 
Closes #213

### CODE REVIEW

File: backend/src/search/SearchIndexBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+  if (typeof value === 'string') {
+    return value;
+  }
+
+  // If the value is an array, convert each element and combine them via flattenArray.
+  if (Array.isArray(value)) {
+    if (value.length > 1) {
+      return flattenArray(
+        <string[]>(
+          value.map((entry) => convertToString(entry)).filter((entry) => entry)
```
Comment (joluj) [inline backend/src/search/SearchIndexBuilder.ts:35]: This will remove empty strings, but I thinks this is also fine right?

File: backend/src/search/SearchIndexBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    }
+
+    return convertToString(value[0]);
+  }
+
+  // Try to invoke a 'toString()' function, that may be present.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  if (typeof (<any>value).toString === 'function') {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const result = (<any>value).toString();
```
Comment (joluj) [inline backend/src/search/SearchIndexBuilder.ts:47]:
```suggestion
  if (typeof value === 'object' && typeof value?.toString === 'function') {
    const result = value.toString();
```

No any

File: backend/src/search/SearchIndexBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    }
+  }
+
+  // The value cannot be stringified.
+  return null;
+}
+
+/**
+ * Translated an edge to an index-entry and records it in an entries collection.
+ * @param node The edge to translate.
```
Comment (joluj) [inline backend/src/search/SearchIndexBuilder.ts:60]:
```suggestion
 * @param edge The edge to translate.
```

File: backend/src/search/SearchIndexBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+  entries.push(nodeEntityEntry);
+
+  recordEntityType(nodeType, 'node-type', entries);
+}
+
+/**
+ * A search index builder that can be used to build a search index from a complete dataset.
+ */
+@Injectable()
+export class SearchIndexBuilder {
```
Comment (joluj) [inline backend/src/search/SearchIndexBuilder.ts:246]: Move the functions above in another file, they take too much space.

File: backend/src/search/SearchIndexBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    const entries: IndexEntry[] = [];
+
+    // TODO: For large databases it would be beneficial, if we read the entries in smaller batches,
+    //       so we don't have to copy the entire dataset into memory (again).
+    //       The problem here, is that we need the property names before adding the entries to the
+    //       index. This can be done by executing an up-front query for the schema of the data,
+    //       as is done in the SchemaService.
+
+    // Get all nodes from the database
+    const nodeResults = await this.neo4jService.read(
```
Comment (joluj) [inline backend/src/search/SearchIndexBuilder.ts:267]: use service

File: backend/src/search/SearchIndexBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+      (record) => record.toObject() as Node
+    );
+
+    // eslint-disable-next-line no-restricted-syntax
+    for (const node of nodes) {
+      recordNode(node, entries);
+    }
+
+    // Get all edges from the database
+    const edgeResults = await this.neo4jService.read(
```
Comment (joluj) [inline backend/src/search/SearchIndexBuilder.ts:281]: use service

File: backend/src/search/SearchIndexBuilder.ts [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+      (record) => record.toObject() as Edge
+    );
+
+    // eslint-disable-next-line no-restricted-syntax
+    for (const edge of edges) {
+      recordEdge(edge, entries);
+    }
+
+    // Get all node types from the database
+    const nodeTypeResults = await this.neo4jService.read(
```
Comment (joluj) [inline backend/src/search/SearchIndexBuilder.ts:299]: use service

File: backend/src/search/SearchIndexBuilder.ts [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+
+    const nodeTypes = parseNeo4jEntityInfo(nodeTypeResults.records, 'node');
+
+    // eslint-disable-next-line no-restricted-syntax
+    for (const nodeType of nodeTypes) {
+      recordNodeType(nodeType, entries);
+    }
+
+    // Get all edge types from the database
+    const edgeTypeResults = await this.neo4jService.read(
```
Comment (joluj) [inline backend/src/search/SearchIndexBuilder.ts:311]: use service

File: frontend/cypress/integration/search.e2e.js [hunk 9]
```diff
@@ -20,37 +20,29 @@ context('Searchbar', () => {
...
-    cy.intercept(`${apiBaseUrl}/getNodesById*`, spyGetNodesById).as(
-      'getNodesQuery'
-    );
 
     // Act 1
     cy.get('.SearchBar').type('keanu');
 
     // Assert 1
-    cy.wait(['@searchQuery', '@getNodesQuery']).then(() => {
+    cy.wait(['@searchQuery']).then(() => {
```
Comment (joluj) [inline frontend/cypress/integration/search.e2e.js:30]: why did you remove all that extra tests?

File: frontend/cypress/unit/search.test.tsx [hunk 10]
```diff
@@ -33,43 +32,29 @@ context('Searchbar', () => {
...
     cy.intercept(`${apiBaseUrl}/search/all*`, spySearch).as('searchQuery');
-    cy.intercept(`${apiBaseUrl}/getNodesById*`, spyGetNodesById).as(
-      'getNodesQuery'
-    );
 
     // Act 1
     cy.get('.SearchBar').type('keanu');
     // Assert 1
-    cy.wait(['@searchQuery', '@getNodesQuery']).then(() => {
+    cy.wait(['@searchQuery']).then(() => {
```
Comment (joluj) [inline frontend/cypress/unit/search.test.tsx:40]: why did you remove that checks?

**Review by joluj — CHANGES_REQUESTED** [review]
Overall very nice, only some little improvements in the normal code.

Why did you remove some extra checks in the tests?

File: frontend/cypress/integration/search.e2e.js (hunk 9 above)
Comment (CatoLeanTruetschel) [reply frontend/cypress/integration/search.e2e.js:30]: Because this is no longer called any longer. This was the reason to even do these changes.

File: frontend/cypress/unit/search.test.tsx (hunk 10 above)
Comment (CatoLeanTruetschel) [reply frontend/cypress/unit/search.test.tsx:40]: Because this is no longer called any longer. This was the reason to even do these changes.

File: backend/src/search/SearchIndexBuilder.ts (hunk 1 above)
Comment (CatoLeanTruetschel) [reply backend/src/search/SearchIndexBuilder.ts:35]: Yes, that's intended.

File: backend/src/search/SearchIndexBuilder.ts (hunk 8 above)
Comment (CatoLeanTruetschel) [reply backend/src/search/SearchIndexBuilder.ts]: Done

File: backend/src/search/SearchIndexBuilder.ts (hunk 7 above)
Comment (CatoLeanTruetschel) [reply backend/src/search/SearchIndexBuilder.ts]: Done

File: backend/src/search/SearchIndexBuilder.ts (hunk 6 above)
Comment (CatoLeanTruetschel) [reply backend/src/search/SearchIndexBuilder.ts]: Done

File: backend/src/search/SearchIndexBuilder.ts (hunk 5 above)
Comment (CatoLeanTruetschel) [reply backend/src/search/SearchIndexBuilder.ts]: Done

File: backend/src/search/SearchIndexBuilder.ts (hunk 4 above)
Comment (CatoLeanTruetschel) [reply backend/src/search/SearchIndexBuilder.ts:10]: I refactored the  recording stuff to a dedicated type and moved the rather generic functions to own files in the utils dir.

File: backend/src/search/SearchIndexBuilder.ts (hunk 3 above)
Comment (CatoLeanTruetschel) [reply backend/src/search/SearchIndexBuilder.ts]: Done

File: backend/src/search/SearchIndexBuilder.ts (hunk 2 above)
Comment (CatoLeanTruetschel) [reply backend/src/search/SearchIndexBuilder.ts]: Good catch! Done

File: backend/src/utils/any.ts [hunk 11]
```diff
@@ -1,4 +1,4 @@
-export default function any<T>(
+export function any<T>(
```
Comment (joluj) [inline backend/src/utils/any.ts:1]: didn't we want to remove that function? (Might be an other issue)

File: backend/src/utils/flattenArray.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Flattens the specified string array.
+ */
+export function flattenArray(array: string[]): string {
```
Comment (joluj) [inline backend/src/utils/flattenArray.ts:4]: maybe a better name would be `stringifyArray`

**Review by joluj — APPROVED** [review]

File: backend/src/utils/any.ts (hunk 11 above)
Comment (CatoLeanTruetschel) [reply backend/src/utils/any.ts:1]: Yes, thats another story. Will be sone in a follow-up

File: backend/src/utils/flattenArray.ts (hunk 12 above)
Comment (CatoLeanTruetschel) [reply backend/src/utils/flattenArray.ts:4]: This is tracked by #235
